### PR TITLE
Marking Invalid Values as Omitted

### PR DESCRIPTION
**Pull request by kevinkreiser**
`speed` , `course` and `horizontalAccuracy` can all hold invalid values. `MBFixLocation` uses `nil` to communicate the omission of an optional value.

Interestingly an invalid `horizontalAccuracy` also means that the `CLLocationCoordinate2D` will also be bunk, but I've not taken any measures there as that is a required parameter by `MBFixLocation`

I've also made the changes in the other direction as well and marked omitted values from `MBFixLocation` as invalid in the constructed `CLLocation`

### CODE REVIEW

File: MapboxCoreNavigation/CLLocation.swift [hunk 1]
```diff
@@ -158,11 +158,4 @@ extension CLLocation {
     func shifted(to newTimestamp: Date) -> CLLocation {
         return CLLocation(coordinate: coordinate, altitude: altitude, horizontalAccuracy: horizontalAccuracy, verticalAccuracy: verticalAccuracy, course: course, speed: speed, timestamp: newTimestamp)
     }
-    
-    convenience init(fixLocation: MBFixLocation) {
```
Comment (kevinkreiser) [inline MapboxCoreNavigation/CLLocation.swift]: this appeared to be duplicated

File: MapboxCoreNavigation/CLLocation.swift [hunk 2]
```diff
@@ -30,10 +30,10 @@ extension CLLocation {
     convenience init(_ location: MBFixLocation) {
         self.init(coordinate: location.coordinate,
                   altitude: location.altitude?.doubleValue ?? 0,
-                  horizontalAccuracy: location.accuracyHorizontal?.doubleValue ?? 0,
+                  horizontalAccuracy: location.accuracyHorizontal?.doubleValue ?? -1,
                   verticalAccuracy: 0,
-                  course: location.bearing?.doubleValue ?? 0,
-                  speed: location.speed?.doubleValue ?? 0,
+                  course: location.bearing?.doubleValue ?? -1,
+                  speed: location.speed?.doubleValue ?? -1,
```
Comment (kevinkreiser) [inline MapboxCoreNavigation/CLLocation.swift]: when they are missing from `MBFixLocation` we mark as invalid in the `CLLocation`

File: MapboxCoreNavigation/MBNavigator.swift [hunk 3]
```diff
@@ -7,10 +7,10 @@ extension MBFixLocation {
     convenience init(_ location: CLLocation) {
         self.init(coordinate: location.coordinate,
                   time: location.timestamp,
-                  speed: location.speed as NSNumber,
-                  bearing: location.course as NSNumber,
+                  speed: location.speed >= 0 ? location.speed as NSNumber : nil,
+                  bearing: location.course >= 0 ? location.course as NSNumber : nil,
                   altitude: location.altitude as NSNumber,
-                  accuracyHorizontal: location.horizontalAccuracy as NSNumber,
+                  accuracyHorizontal: location.horizontalAccuracy >= 0 ? location.horizontalAccuracy as NSNumber : nil,
```
Comment (kevinkreiser) [inline MapboxCoreNavigation/MBNavigator.swift]: when they are invalid in the `CLLocation` we mark them as omitted in the `MBFixLocation`

File: MapboxCoreNavigation/MBNavigator.swift [hunk 4]
```diff
@@ -7,10 +7,10 @@ extension MBFixLocation {
     convenience init(_ location: CLLocation) {
         self.init(coordinate: location.coordinate,
                   time: location.timestamp,
-                  speed: location.speed as NSNumber,
-                  bearing: location.course as NSNumber,
+                  speed: location.speed >= 0 ? location.speed as NSNumber : nil,
+                  bearing: location.course >= 0 ? location.course as NSNumber : nil,
                   altitude: location.altitude as NSNumber,
```
Comment (frederoni) [inline MapboxCoreNavigation/MBNavigator.swift]: altitude is nullable but supports negative values to indicate below sea level so this conversion is correct

Comment (kevinkreiser) [reply MapboxCoreNavigation/MBNavigator.swift]: yep that was my thoughts exactly. we want to have correct altitudes when passing through death valley or doing u-boot routing ;)

**Pull request merged**